packages/runtime-core/src/componentEmits.ts

```diff
@@ -151,10 +152,14 @@ export function emit(
   }
 
   let args = rawArgs
-  const isModelListener = event.startsWith('update:')
+  const isModelListener =
+    __COMPAT__ && isCompatEnabled(DeprecationTypes.COMPONENT_V_MODEL, instance)
+      ? compatModelEventPrefix + event in props
+      : event.startsWith('update:')
 
   // for v-model update:xxx events, apply modifiers on args
-  const modifiers = isModelListener && getModelModifiers(props, event.slice(7))
+  const modifiers =
+    isModelListener && getModelModifiers(instance, props, event.slice(7))
```

Comment:
The third argument here is not used when `__COMPAT__ && isCompatEnabled(DeprecationTypes.COMPONENT_V_MODEL, instance)`, but adding a ternary for that here doesn't seem worth it.

Reply:
```js
let isModelListener
let modifiers
if(
  __COMPAT__ && 
  isCompatEnabled(DeprecationTypes.COMPONENT_V_MODEL, instance) && 
  (isModelListerner = (compatModelEventPrefix + event in props))
){  
  modifiers = props.modelModifiers
}
else if(isModelListener = event.startsWith('update:')){
  modifiers = getModelModifiers(props, event.slice(7))
}
```
How about this?
- `isCompatEnabled(DeprecationTypes.COMPONENT_V_MODEL, instance)` will only be checked once.
- `getModelModifiers` also does not need any changes. The line `const modifiers = getModelModifiers(i, props, camelizedName)` in `useModel` does not need any changes either, because `defineModel` is not supported in the Compat mode.

Reply:
Sounds good! I've pushed a commit with this. In fact, I've pushed another, which simplifies this a bit more. I think that `compatModelEventPrefix + event in props` can only be `true` if `isCompatEnabled(DeprecationTypes.COMPONENT_V_MODEL, instance)` anyway, which is done in [`convertLegacyVModelProps`](https://github.com/vuejs/core/blob/9d1a0b2ef0d14d94ed1619669c9cb628bb6eb256/packages/runtime-core/src/compat/componentVModel.ts#L16-L53).

Thoughts?